scheduler/pkg/kafka/pipeline/kafkamanager.go

```diff
@@ -197,14 +215,52 @@ func (km *KafkaManager) Infer(
 	requestId string,
 ) (*Request, error) {
 	logger := km.logger.WithField("func", "Infer")
-	km.mu.RLock()
-	pipeline, err := km.loadOrStorePipeline(resourceName, isModel)
+
+	var (
+		pipeline *Pipeline
+		err      error
+	)
+	pipeline, err = km.LoadPipeline(resourceName, isModel)
 	if err != nil {
-		km.mu.RUnlock()
-		return nil, err
+		if isModel {
+			// We only allow lazy loading of model. This path should not be reached
+			// through envoy but only sending requests directly to the pipeline gateway.
+			// Note that due to lazy loading, having multiple replicas of the pipelinegw
+			// may result in unexpected behaviour. This is because due to the load balancing
+			// of the requests, consumers with the same group id may end up consuming from
+			// different topics.
+			logger.Warning("Lazy loading of the models is only suppoerted for a single pipeline gateway instance.")
+			err := km.StorePipeline(resourceName, true)
+			if err != nil {
+				return nil, fmt.Errorf("failed to store model pipeline %s: %w", resourceName, err)
+			}
+
+			// In this case, the existence of the pipeline is guaranteed
+			pipeline, _ = km.LoadPipeline(resourceName, true)
+		} else {
+			return nil, err
+		}
 	}
+
+	// We lock here so we don't reasign the partitions while producing a message.
+	km.mu.RLock()
+	defer km.mu.RUnlock()
+
+	partitions := pipeline.consumer.partitions
+	if len(partitions) == 0 {
+		return nil, fmt.Errorf("no partitions assigned for topic %s", resourceName)
+	}
+
+	// Randomly select a partition to produce the message to
+	partition := partitions[rand.Intn(len(partitions))]
+	logger.Debugf("Using partition %d for resource %s", partition, resourceName)
+
 	// Use composite key to differentiate multiple pipelines (i.e. mirror) using the same message
-	compositeKey := getCompositeKey(resourceName, requestId, ".")
+	// Note that we add the partition to the key to ensure that the message will be sent to
+	// a partition for which the consumer is subscribed. For modelgw, it is enought to send the
+	// message to the same partition as the one we read from. For dataflow engine on the other hand,
+	// we need to read the partition from the request id.
+	compositeKey := getCompositeKey(strconv.Itoa(int(partition)), resourceName, requestId, ".")
```

Comment:
This is an elegant way out of the fundamental problem of each pipeline-gw needing to be both a producer into a partition and a consumer from the same exact partition, whith dataflow-engine controlling this latter partition no.